client/state/sites/reducer.js

```diff
@@ -98,18 +99,38 @@ export function items( state = {}, action ) {
 			return state;
 		}
 
-		case SITE_RECEIVE: {
-			const site = pick( action.site, VALID_SITE_KEYS );
-			return Object.assign( {}, state, {
-				[ site.ID ]: site
-			} );
-		}
-
+		case SITE_RECEIVE:
 		case SITES_RECEIVE:
-			return action.sites.reduce( ( memo, site ) => {
-				memo[ site.ID ] = pick( site, VALID_SITE_KEYS );
+		case SITES_UPDATE:
+			// Normalize incoming site(s) to array
+			const sites = action.site ? [ action.site ] : action.sites;
+
+			// SITES_RECEIVE occurs when we receive the entire set of user
+			// sites (replace existing state). Otherwise merge into state.
+			const initialMemo = SITES_RECEIVE === action.type ? {} : state;
+
+			return reduce( sites, ( memo, site ) => {
+				// If we're not already tracking the site upon an update, don't
+				// merge into state (TODO: Consider dropping condition once
+				// sites-list abolished)
+				if ( SITES_UPDATE === action.type && ! memo[ site.ID ] ) {
+					return memo;
+				}
+
+				// Bypass if site object hasn't change
+				const transformedSite = pick( site, VALID_SITE_KEYS );
+				if ( isEqual( memo[ site.ID ], transformedSite ) ) {
+					return memo;
+				}
+
+				// Avoid mutating state
+				if ( memo === state ) {
+					memo = { ...state };
+				}
+
+				memo[ site.ID ] = transformedSite;
 				return memo;
-			}, {} );
+			}, initialMemo );
```

Comment:
>This is looking good, but I find the term `memo` confusing. it looks like's it's being used to represent the previous sites state. did the term come from "memoization" or "memory"?

I've used for some time, but if I were to guess, I might have picked it up from Underscore.js's documentation back in the day: http://underscorejs.org/#reduce

Doesn't seem like [too uncommon a term](https://www.google.com/search?q=array+reduce+memo) and you're probably right it's a convenient derivation of one of either memory or memoization.

What would your suggested alternative be?